File: packages/ts/src/fingerprint.ts
/** Top-level orchestration: walk a repo and produce a full DetectionReport. */
import { resolve } from "node:path";
import type { DetectionReport, GeneratedBy } from "./types.js";
import { loadMatrix } from "./matrix.js";
import { walk } from "./walker.js";
import { collectSignals } from "./signals.js";
import { parseManifests } from "./parsers.js";
import { matchFrameworks, matchTesting } from "./frameworks.js";
import { inferBuildTools, inferPackageManagers, resolveTopology } from "./topology.js";
import { resolveInfrastructure } from "./infra.js";
import { aggregate, assembleReport, computeSubRepos } from "./report.js";
import { primaryManifestCount } from "./confidence.js";

export interface FingerprintOptions {
  generatedBy?: GeneratedBy;
  /** Timestamp override (mostly for deterministic tests). */
  now?: string;
  /** Deep (shadow) scan: monorepo-aware dominance fallback + sub-repo enumeration. */
  deep?: boolean;
}

export function fingerprint(root: string, opts: FingerprintOptions = {}): DetectionReport {
  const absRoot = resolve(root);
  const matrix = loadMatrix();
  const tree = walk(absRoot);
  const deep = opts.deep ?? false;

  const signals = collectSignals(matrix, tree);
  const pools = parseManifests(absRoot, tree.files);
  const { ecosystems, dominantEcosystem } = aggregate(matrix, signals, "confidence", deep);

  let topology = resolveTopology(matrix, absRoot, tree.files);
  let subRepos;
  if (deep) {
    subRepos = computeSubRepos(signals);
    // Deep topology inference: >= 2 sub-repos, no root-proximate primary manifest, and no
    // workspace/monorepo marker already detected => a marker-less "multi-repo" monorepo.
    const rootPrimaries = primaryManifestCount(signals, true);
    if (subRepos.length >= 2 && rootPrimaries === 0 && topology.type === "single") {
      topology = { ...topology, type: "monorepo", tool: null };
    }
  }

  return assembleReport({
    root: absRoot,
    generatedBy: opts.generatedBy ?? "ts",
    generatedAt: opts.now ?? new Date().toISOString(),
    ecosystems,
    packageManagers: inferPackageManagers(matrix, tree.files),
    buildTools: inferBuildTools(matrix, tree.files),
    topology,
    frameworks: matchFrameworks(matrix, pools),
    testing: matchTesting(matrix, pools),
    infrastructure: resolveInfrastructure(matrix, tree.files),
    dominantEcosystem,
    ...(subRepos ? { subRepos } : {}),
  });
}

File: packages/py/src/repo_fingerprint/fingerprint.py
"""Top-level orchestration: walk a repo and produce a full DetectionReport."""
from __future__ import annotations

import os
from datetime import datetime, timezone

from .confidence import primary_manifest_count
from .frameworks import match_frameworks, match_testing
from .infra import resolve_infrastructure
from .matrix import load_matrix
from .parsers import parse_manifests
from .report import aggregate, assemble_report, compute_sub_repos
from .signals import collect_signals
from .topology import infer_build_tools, infer_package_managers, resolve_topology
from .types import DetectionReport, Topology
from .walker import walk


def fingerprint(
    root: str, generated_by: str = "py", now: str | None = None, deep: bool = False
) -> DetectionReport:
    abs_root = os.path.abspath(root)
    matrix = load_matrix()
    tree = walk(abs_root)

    signals = collect_signals(matrix, tree)
    pools = parse_manifests(abs_root, tree.files)
    ecosystems, dominant = aggregate(matrix, signals, "confidence", deep)

    topology = resolve_topology(matrix, abs_root, tree.files)
    sub_repos = None
    if deep:
        sub_repos = compute_sub_repos(signals)
        # Deep topology inference: >= 2 sub-repos, no root-proximate primary manifest, and no
        # workspace/monorepo marker already detected => a marker-less "multi-repo" monorepo.
        root_primaries = primary_manifest_count(signals, True)
        if len(sub_repos) >= 2 and root_primaries == 0 and topology.type == "single":
            topology = Topology(type="monorepo", tool=None, signals=topology.signals)

    generated_at = now or datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    return assemble_report(
        root=abs_root,
        generated_by=generated_by,
        generated_at=generated_at,
        ecosystems=ecosystems,
        package_managers=infer_package_managers(matrix, tree.files),
        build_tools=infer_build_tools(matrix, tree.files),
        topology=topology,
        frameworks=match_frameworks(matrix, pools),
        testing=match_testing(matrix, pools),
        infrastructure=resolve_infrastructure(matrix, tree.files),
        dominant_ecosystem=dominant,
        sub_repos=sub_repos,
    )
